### PR TITLE
DBAASCLI-69 Add operator deployment check

### PR DESCRIPTION
**Pull request by nonemax**

### CODE REVIEW

**Comment by JNKPercona** [comment]
License check is ok.

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
License check is ok.

**Comment by JNKPercona** [comment]
License check is ok.

File: dbaas-lib/engines/k8s-psmdb/psmdb.go [hunk 1]
```diff
@@ -131,17 +129,27 @@ func NewPSMDBController(envCrt, provider string) (*PSMDB, error) {
...
-
-	for i, o := range v[defaultVersion].k8s.Bundle {
-		if o.Kind == "Deployment" && o.Name == p.operatorName() {
-			v[defaultVersion].k8s.Bundle[i].Data = strings.Replace(o.Data, "{{image}}", operatorVersion, -1)
+	switch i := len(version); {
+	case i == 0:
+		version = defaultVersion
+	default:
+		if _, ok := objects[version]; !ok {
+			return errors.Errorf("unsupporeted version %s", version)
```
Comment (dAdAbird) [inline dbaas-lib/engines/k8s-psmdb/psmdb.go]:
```go
	switch len(version) {
	case 0:
...
```

but I'd rather go for
```go
	if len(version) == 0 {
		version = defaultVersion
	}
	if _, ok := objects[version]; !ok {
		return errors.Errorf("unsupporeted version %s", version)
	}
```
instead of this switch

File: dbaas-lib/k8s/cmd.go [hunk 2]
```diff
@@ -132,18 +136,45 @@ func (p Cmd) readOperatorLogs(operatorName string) ([]byte, error) {
...
+		data, err = p.runCmd(p.execCommand, "get", typ, name, "-n", p.Namespace, "-o=jsonpath={"+jsonPath+"}")
+	default:
+		data, err = p.runCmd(p.execCommand, "get", typ, name, "-o=jsonpath={"+jsonPath+"}")
+	}
+	if err != nil && strings.Contains(err.Error(), "not found") {
+		err = ErrNotFound
+	}
+
+	return
+}
```
Comment (dAdAbird) [inline dbaas-lib/k8s/cmd.go]:
```suggestion
func (p Cmd) GetObjectsElement(typ, name, jsonPath string) ([]byte, error) {
	args := []string{p.execCommand, "get", typ, name, "-o=jsonpath={" + jsonPath + "}"}
	if len(p.Namespace) > 0 {
		args = append(args, []string{"-n", p.Namespace}...)
	}
	data, err = p.runCmd(p.execCommand, args...)
	if err != nil && strings.Contains(err.Error(), "not found") {
		err = ErrNotFound
	}

	return data, err
}
```

Comment (dAdAbird) [reply dbaas-lib/k8s/cmd.go]: Are you sure "not found" gonna be in the error not in data? `p.runCmd` basically runs `exec.Cmd().CombinedOutput()` which returns both stdout and stderr as the first arg. And the err is for CombinedOutput's errors, not stderr'

File: dbaas-lib/k8s/cmd.go [hunk 3]
```diff
@@ -160,9 +191,25 @@ func (p Cmd) DeleteObject(typ, name string) error {
...
+	defer os.Remove(fileName)
+
+	_, err = f.Write([]byte(k8sObj))
+	if err != nil {
+		return errors.Wrap(err, "write cr file")
+	}
+	if len(namespace) > 0 {
+		_, err = p.runCmd(p.execCommand, "apply", namespace, "-f", fileName)
+	} else {
+		_, err = p.runCmd(p.execCommand, "apply", "-f", fileName)
```
Comment (dAdAbird) [inline dbaas-lib/k8s/cmd.go]: 1. Why you need a tmp file? Why the old method doesn't work anymore?
2. In case you need it, use https://golang.org/pkg/io/ioutil/#TempFile and maybe https://golang.org/pkg/os/#TempDir. But no guarantees that tmp dir exists and have accessible permissions. So you have to check it as well.

File: dbaas-lib/k8s/check.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+import (
+	"fmt"
+	"strings"
+
+	v "github.com/hashicorp/go-version"
+	"github.com/pkg/errors"
+)
+
+func (p *Cmd) PreCheck(name, version, operatorName, operatorImage, objectName string, supportedVersions map[string]string) (warnings []string, errArr []error) {
```
Comment (dAdAbird) [inline dbaas-lib/k8s/check.go]: https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters

> Naked returns are okay if the function is a handful of lines. Once it's a medium sized function, be explicit with your return values.

it's really hard to follow what this method returns

Comment (dAdAbird) [inline dbaas-lib/k8s/check.go]: Do we have to have errors slice?
As far as I can see you never return more than one error

File: dbaas-lib/k8s/check.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+	}
+
+	i := oVersion.Compare(dVersion)
+	switch i {
+	case 0:
+		return
+	default:
+		warnings = append(warnings, fmt.Sprintf("%s operator with version %s already installed. Trying work with it", objectName, deployedVersion))
+	}
+	return
```
Comment (dAdAbird) [inline dbaas-lib/k8s/check.go]:
```go
	if oVersion.Compare(dVersion) != 0 {
		warnings = append(warnings, fmt.Sprintf("%s operator with version %s already installed. Trying work with it", objectName, deployedVersion))
	}
	return
```

File: dbaas-lib/k8s/cmd.go [hunk 6]
```diff
@@ -132,18 +136,45 @@ func (p Cmd) readOperatorLogs(operatorName string) ([]byte, error) {
...
+		data, err = p.runCmd(p.execCommand, "get", typ)
+	}
+
+	if err != nil && strings.Contains(err.Error(), "not found") {
+		err = ErrNotFound
+	} else if err != nil && strings.Contains(string(data), "doesn't have a resource") {
+		err = ErrNotFound
+	}
+	if strings.Contains(string(data), "No resources found") {
+		err = ErrNotFound
```
Comment (dAdAbird) [inline dbaas-lib/k8s/cmd.go]: see comments on GetObjectsElement()
and don't overuse "switch" when you can go for "if"

File: dbaas-cli/cmd/mongo/create.go [hunk 7]
```diff
@@ -72,6 +72,16 @@ var createCmd = &cobra.Command{
 			Provider:      *provider,
 			RootPass:      *rootPass,
 		}
+		warns, errArr := dbaas.PreCheck(instance)
+		if len(errArr) > 0 {
+			for _, e := range errArr {
+				log.Println(e)
```
Comment (dAdAbird) [inline dbaas-cli/cmd/mongo/create.go]:
```suggestion
				log.Println("Error:", e)
```

File: dbaas-cli/cmd/mongo/create.go [hunk 8]
```diff
@@ -72,6 +72,16 @@ var createCmd = &cobra.Command{
...
+			}
+			return
+		}
+		for _, w := range warns {
+			log.Println(w)
```
Comment (dAdAbird) [inline dbaas-cli/cmd/mongo/create.go]:
```suggestion
			log.Println("Warning:", w)
```

File: dbaas-cli/cmd/mongo/create.go [hunk 9]
```diff
@@ -72,6 +72,16 @@ var createCmd = &cobra.Command{
...
+		}
```
Comment (dAdAbird) [inline dbaas-cli/cmd/mongo/create.go:78]: Maybe it worth to show Warnings as well even there are some errors

File: dbaas-cli/cmd/mysql/create.go [hunk 10]
```diff
@@ -73,6 +73,17 @@ var createCmd = &cobra.Command{
...
+		/*warns, errArr := dbaas.PreCheck(instance)
+		if len(errArr) > 0 {
+			for _, e := range errArr {
+				log.Println(e)
+			}
+			return
+		}
+		for _, w := range warns {
+			log.Println(w)
+		}*/
```
Comment (dAdAbird) [inline dbaas-cli/cmd/mysql/create.go]: ?

File: dbaas-cli/cmd/mysql/delete.go [hunk 11]
```diff
@@ -86,13 +86,22 @@ var delCmd = &cobra.Command{
...
+		/*warns, errArr := dbaas.PreCheck(instance)
+		if len(errArr) > 0 {
+			for _, e := range errArr {
+				log.Println(e)
+			}
+			return
+		}
+		for _, w := range warns {
+			log.Println(w)
+		}*/
```
Comment (dAdAbird) [inline dbaas-cli/cmd/mysql/delete.go]: ?

File: dbaas-lib/engines/k8s-psmdb/options.go [hunk 12]
```diff
@@ -7,13 +7,13 @@ import (
...
-	c.SetDefaults()
-	err := options.Parse(&c, reflect.TypeOf(c), opts)
+	//c := objects[defaultVersion].psmdb
+	p.conf.SetDefaults()
+	err := options.Parse(&p.conf, reflect.TypeOf(p.conf), opts)
 	if err != nil {
 		return err
 	}
-	p.conf = c
+	//	p.conf = c
```
Comment (dAdAbird) [inline dbaas-lib/engines/k8s-psmdb/options.go]: ?

File: dbaas-lib/engines/k8s-psmdb/options.go [hunk 13]
```diff
@@ -7,13 +7,13 @@ import (
 )
 
 func (p *PSMDB) ParseOptions(opts string) error {
-	c := objects[defaultVersion].psmdb
-	c.SetDefaults()
-	err := options.Parse(&c, reflect.TypeOf(c), opts)
+	//c := objects[defaultVersion].psmdb
```
Comment (dAdAbird) [inline dbaas-lib/engines/k8s-psmdb/options.go]: ?

File: dbaas-lib/engines/k8s-psmdb/psmdb.go [hunk 14]
```diff
@@ -15,12 +14,12 @@ import (
 )
 
 const (
-	provider               = "k8s"
-	engine                 = "psmdb"
-	defaultVersion Version = "1.3.0"
+	provider       = "k8s"
+	engine         = "psmdb"
+	defaultVersion = "1.3.0"
```
Comment (dAdAbird) [inline dbaas-lib/engines/k8s-psmdb/psmdb.go:19]: Why did you drop Version type?

File: dbaas-lib/k8s/check.go (hunk 4 above)
Comment (dAdAbird) [reply dbaas-lib/k8s/check.go]: Oh, I see one place. But, according to logic, there are no multiple errors. Either you return some technical error (e.g. supported version, can't parse version etc.) or "no operator + object exists" - in fact, a single error.

**Review by dAdAbird — CHANGES_REQUESTED** [review]
In addition, it's nice to have some integration test that covers introduced cases.

**Comment by JNKPercona** [comment]
License check is ok.

File: dbaas-cli/integtests/k8s-cmd.go [hunk 15]
```diff
@@ -32,3 +32,11 @@ func GetK8SObject(typ, name string) ([]byte, error) {
 	}
 	return []byte(o), nil
 }
+
+func DeleteDeployment(name string) ([]byte, error) {
+	o, err := runCmd("kubectl", "delete", "deployment", name)
+	if err != nil {
+		return nil, err
+	}
+	return []byte(o), nil
```
Comment (dAdAbird) [inline dbaas-cli/integtests/k8s-cmd.go]: The only user of this parameter cast it back to a string. So maybe just return string here?

File: dbaas-lib/engines/k8s-psmdb/psmdb.go (hunk 14 above)
Comment (dAdAbird) [reply dbaas-lib/engines/k8s-psmdb/psmdb.go:19]: @nonemax ^^^

File: dbaas-lib/k8s/check.go [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+		return warnings, errors.Errorf("no operator but existing %s objects", objectName)
+	}
+
+	if len(deploymentImage) == 0 {
+		return warnings, err
+	}
+
+	if string(deploymentImage) == operatorImage {
+		return warnings, nil
+	}
```
Comment (dAdAbird) [inline dbaas-lib/k8s/check.go]: `err` always will be `nil` if this code reached.
So `return warnings, err` and then `return warnings, nil` looks confusing.
What should be done if `deploymentImage` is empty?

File: dbaas-cli/integtests/k8s-psmdb.go [hunk 17]
```diff
@@ -233,3 +242,21 @@ func (psmdb *KuberPSMDB) DeleteDB(preserve bool) error {
 	}
 	return nil
 }
+
+func (psmdb *KuberPSMDB) DeleteDBWithoutOperator() error {
+	fmt.Println("Run delete-db " + psmdb.dbName + ". Without Operator")
+	_, err := DeleteDeployment("percona-server-mongodb-operator")
+	if err != nil {
+		return errors.Wrap(err, "run delete-db cmd")
```
Comment (dAdAbird) [inline dbaas-cli/integtests/k8s-psmdb.go]:
```suggestion
		return errors.Wrap(err, "delete operator deployment")
```

File: dbaas-lib/k8s/cmd.go [hunk 18]
```diff
@@ -140,10 +160,25 @@ func (p Cmd) GetObject(typ, name string) ([]byte, error) {
...
+	} else if err != nil && strings.Contains(string(data), "doesn't have a resource") {
+		err = ErrNotFound
+	}
+
+	if strings.Contains(string(data), "No resources found") {
+		err = ErrNotFound
+	}
+
+	return data, err
```
Comment (dAdAbird) [inline dbaas-lib/k8s/cmd.go]: `p.runCmd` at the end of the day runs https://pkg.go.dev/os/exec?tab=doc#Cmd.CombinedOutput so there is no way some k8s error may end up in the `err`. `err` could be not nil only if os/exec failed.

File: dbaas-lib/k8s/cmd.go [hunk 19]
```diff
@@ -160,9 +195,25 @@ func (p Cmd) DeleteObject(typ, name string) error {
...
+	if err != nil {
+		return errors.Wrap(err, "write cr file")
+	}
+	if len(namespace) > 0 {
+		_, err = p.runCmd(p.execCommand, "apply", namespace, "-f", fileName)
+	} else {
+		_, err = p.runCmd(p.execCommand, "apply", "-f", fileName)
 	}
-	_, err := p.runCmd("sh", "-c", "cat <<-EOF | "+p.execCommand+" apply "+namespace+"-f -\n"+k8sObj+"\nEOF")
+	//_, err = p.runCmd("sh", "-c", "cat <<-EOF | "+p.execCommand+" apply "+namespace+"-f -\n"+k8sObj+"\nEOF")
```
Comment (dAdAbird) [inline dbaas-lib/k8s/cmd.go]: do we need this?

**Review by dAdAbird — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
License check is ok.

File: dbaas-lib/k8s/cmd.go (hunk 19 above)
Comment (nonemax) [reply dbaas-lib/k8s/cmd.go]: https://github.com/Percona-Lab/percona-dbaas-cli/pull/100/commits/3f844df711bf1b35c32ec89de8851e67756b6292

File: dbaas-cli/integtests/k8s-psmdb.go (hunk 17 above)
Comment (nonemax) [reply dbaas-cli/integtests/k8s-psmdb.go]: https://github.com/Percona-Lab/percona-dbaas-cli/pull/100/commits/3f844df711bf1b35c32ec89de8851e67756b6292

File: dbaas-lib/k8s/check.go (hunk 16 above)
Comment (nonemax) [reply dbaas-lib/k8s/check.go]: https://github.com/Percona-Lab/percona-dbaas-cli/pull/100/commits/3f844df711bf1b35c32ec89de8851e67756b6292

File: dbaas-cli/integtests/k8s-cmd.go (hunk 15 above)
Comment (nonemax) [reply dbaas-cli/integtests/k8s-cmd.go]: https://github.com/Percona-Lab/percona-dbaas-cli/pull/100/commits/3f844df711bf1b35c32ec89de8851e67756b6292

File: dbaas-lib/engines/k8s-psmdb/psmdb.go (hunk 14 above)
Comment (nonemax) [reply dbaas-lib/engines/k8s-psmdb/psmdb.go:19]: https://github.com/Percona-Lab/percona-dbaas-cli/pull/100/commits/3f844df711bf1b35c32ec89de8851e67756b6292

File: dbaas-lib/k8s/cmd.go (hunk 18 above)
Comment (dAdAbird) [reply dbaas-lib/k8s/cmd.go]: > err could be not nil only if os/exec failed.

statement is incorrect. I mean `err` can't contain anything from stderr. But err won't be nil if exec failed

**Review by dAdAbird — APPROVED** [review]
